url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/FeederAgent.java

```diff
@@ -0,0 +1,308 @@
+			}
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgRemoveBinDone() {
+		print("Gantry has removed bin from feeder");
+		state = FeederStatus.IDLE;
+		stateChanged();
+	}
+
+	@Override
+	public void msgReceiveBinDone(Bin bin) {
+		print("Received msgReceiveBinDone from graphics");
+		animation.release();
+	}
+
+	@Override
+	public void msgPurgeBinDone(Bin bin) {
+		print("Received msgPurgeBinDone from graphics");
+		state = FeederStatus.IDLE;
+		animation.release();
+	}
+
+	@Override
+	public void msgFlipDiverterDone() {
+		print("Received msgFlipInverterDone from graphics");
+		animation.release();
+	}
+	
+	public void msgThisLanePurged(LaneAgent lane) {
+		for(MyLane currentLane : lanes) {
+			if(currentLane.lane == lane) {
+				currentLane.numPartsNeeded = 0;
+				currentLane.type = null;
+				break;
+			}
+		}
+	}
+
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		// / print("In the scheduler");
+		// synchronized(lanes) {
+		if (state == FeederStatus.IDLE) {
+			synchronized (lanes) {
+				for (MyLane lane : lanes) {
+					if (lane.state == LaneStatus.NEEDS_PARTS) {
+						getParts(lane);
+						return true;
+					}
+				}
+			}
+		}
+		if (state == FeederStatus.RECEIVING_BIN) {
+			receiveBin();
+			return true;
+		}
+		if (state == FeederStatus.FEEDING_PARTS) {
+			synchronized (lanes) {
+				for (MyLane lane : lanes) {
+					if (lane.state == LaneStatus.GIVING_PARTS) {
+						print("Giving parts to lane " + lane.type.getName());
+						if (lanes.get(currentOrientation).equals(lane)) {
+							givePart(lane);
+							return true;
+						} else {
+							// print("Flipping Diverter");
+							flipDiverter();
+							return true;
+						}
+					}
+				}
+			}
+			if(!doesLaneNeedParts()) {
+				state = FeederStatus.PURGING;
+				return true;
+			}
+		}
+		// }
+		if (state == FeederStatus.PURGING) {
+			purgeBin();
+		}
+		return false;
+	}
+	
+	public boolean doesLaneNeedParts() {
+		synchronized(lanes) {
+		for(MyLane lane : lanes) {
+			if(bin != null) {
+			if(lane.type == bin.part.type) {
+				return true;
+			}
+			}
+		}
+		}
+		return false;
+	}
+
+	public void getParts(final MyLane lane) {
+		print("Telling gantry that I needs parts");
+		if (gantry != null) {
+			gantry.msgINeedParts(lane.type, this);
+		}
+		state = FeederStatus.REQUESTED_PARTS;
+		stateChanged();
+	}
+
+	public void givePart(MyLane lane) {
+		print("Giving lane a part");
+		lane.numPartsNeeded--;
+		//Instead of new Part, call GUI method that will initialize the part.
+		lane.lane.msgHereIsPart(new Part(feederGUI.createPartGraphics()));
+		if (lane.numPartsNeeded == 0) {
+			print("shows up when lane " + lane.type.getName()
+					+ " does not need more parts");
+			state = FeederStatus.PURGING;
+			lane.state = LaneStatus.DOES_NOT_NEED_PARTS;
+		}
+		stateChanged();
+	}
+
+	public void purgeBin() {
+		print("Purging this feeder");
+		feederGUI.purgeBin(bin.binGraphics);
+		try {
+			animation.acquire();
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)